In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv('c:\\Users\\samue\\Downloads\\amazon_sales_dataset.csv')

In [4]:
df.count()

order_id            50000
order_date          50000
product_id          50000
product_category    50000
price               50000
discount_percent    50000
quantity_sold       50000
customer_region     50000
payment_method      50000
rating              50000
review_count        50000
discounted_price    50000
total_revenue       50000
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          50000 non-null  int64  
 1   order_date        50000 non-null  object 
 2   product_id        50000 non-null  int64  
 3   product_category  50000 non-null  object 
 4   price             50000 non-null  float64
 5   discount_percent  50000 non-null  int64  
 6   quantity_sold     50000 non-null  int64  
 7   customer_region   50000 non-null  object 
 8   payment_method    50000 non-null  object 
 9   rating            50000 non-null  float64
 10  review_count      50000 non-null  int64  
 11  discounted_price  50000 non-null  float64
 12  total_revenue     50000 non-null  float64
dtypes: float64(4), int64(5), object(4)
memory usage: 5.0+ MB


# 
- change order_date to DATE datatype



In [6]:
df.describe()

order_id    product_id         price  discount_percent  \
count  50000.000000  50000.000000  50000.000000      50000.000000   
mean   25000.500000   2986.848740    252.507260         13.340700   
std    14433.901067   1156.374535    143.025544          9.850694   
min        1.000000   1000.000000      5.010000          0.000000   
25%    12500.750000   1983.000000    127.840000          5.000000   
50%    25000.500000   2983.000000    252.970000         10.000000   
75%    37500.250000   3989.000000    376.335000         20.000000   
max    50000.000000   4999.000000    499.990000         30.000000   

       quantity_sold        rating  review_count  discounted_price  \
count   50000.000000  50000.000000  50000.000000      50000.000000   
mean        2.999400      2.996316    249.329280        218.886566   
std         1.415401      1.154295    144.251981        127.317681   
min         1.000000      1.000000      0.000000          3.530000   
25%         2.000000      2.000000    125.000000        109.680000   
50%         3.000000      3.000000    250.000000        215.805000   
75%         4.000000      4.000000    374.000000        322.702500   
max         5.000000      5.000000    499.000000        499.910000   

       total_revenue  
count   50000.000000  
mean      657.331475  
std       526.223968  
min         3.840000  
25%       240.320000  
50%       505.410000  
75%       968.970000  
max      2499.550000

In [7]:
df.head()

order_id  order_date  product_id product_category   price  \
0         1  2022-04-13        2637            Books  128.75   
1         2  2023-03-12        2300          Fashion  302.60   
2         3  2022-09-28        3670           Sports  495.80   
3         4  2022-04-17        2522            Books  371.95   
4         5  2022-03-13        1717           Beauty  201.68   

   discount_percent  quantity_sold customer_region payment_method  rating  \
0                10              4   North America            UPI     3.5   
1                20              5            Asia    Credit Card     3.7   
2                20              2          Europe            UPI     4.4   
3                15              4     Middle East            UPI     5.0   
4                 0              4     Middle East            UPI     4.6   

   review_count  discounted_price  total_revenue  
0           443            115.88         463.52  
1           475            242.08        1210.40  
2           183            396.64         793.28  
3           212            316.16        1264.64  
4           308            201.68         806.72

#### Date Type Transformation

In [8]:
df['order_date'] = pd.to_datetime(df['order_date'])

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   order_id          50000 non-null  int64         
 1   order_date        50000 non-null  datetime64[ns]
 2   product_id        50000 non-null  int64         
 3   product_category  50000 non-null  object        
 4   price             50000 non-null  float64       
 5   discount_percent  50000 non-null  int64         
 6   quantity_sold     50000 non-null  int64         
 7   customer_region   50000 non-null  object        
 8   payment_method    50000 non-null  object        
 9   rating            50000 non-null  float64       
 10  review_count      50000 non-null  int64         
 11  discounted_price  50000 non-null  float64       
 12  total_revenue     50000 non-null  float64       
dtypes: datetime64[ns](1), float64(4), int64(5), object(3)
memory usage: 5.0+ MB


In [10]:
df["order_date"].min(), df["order_date"].max()

(Timestamp('2022-01-01 00:00:00'), Timestamp('2023-12-31 00:00:00'))

### Step - 3

In [11]:
df["quantity_sold"].describe()

count    50000.000000
mean         2.999400
std          1.415401
min          1.000000
25%          2.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: quantity_sold, dtype: float64

In [12]:
df[df["quantity_sold"] <= 0].shape

(0, 13)

### Step - 4

In [13]:
df["year_week"] = df["order_date"].dt.to_period("W").astype(str)

In [14]:
df[["order_date", "year_week"]].head()

order_date              year_week
0 2022-04-13  2022-04-11/2022-04-17
1 2023-03-12  2023-03-06/2023-03-12
2 2022-09-28  2022-09-26/2022-10-02
3 2022-04-17  2022-04-11/2022-04-17
4 2022-03-13  2022-03-07/2022-03-13

In [15]:
df["year_week"].nunique()

105

### Step 5 — Aggregate to Weekly SKU Demand

In [19]:
weekly_df = (
    df
    .groupby(["year_week", "product_id"], as_index=False)
    .agg({
        "quantity_sold": "sum",
        "price": "mean",
        "discount_percent": "mean",
        'total_revenue': 'sum'
    })
)

In [20]:
weekly_df.rename(columns={
    "quantity_sold": "weekly_quantity_sold", 
    "price": "weekly_avg_price", 
    "discount_percent": "weekly_avg_discount_percent",
}, inplace=True)

In [ ]:
weekly_df

year_week  product_id  weekly_quantity_sold  \
0      2021-12-27/2022-01-02        1042                     4   
1      2021-12-27/2022-01-02        1050                     3   
2      2021-12-27/2022-01-02        1080                     5   
3      2021-12-27/2022-01-02        1094                     3   
4      2021-12-27/2022-01-02        1140                     3   
...                      ...         ...                   ...   
47146  2023-12-25/2023-12-31        4960                     1   
47147  2023-12-25/2023-12-31        4964                     1   
47148  2023-12-25/2023-12-31        4977                     2   
47149  2023-12-25/2023-12-31        4984                     1   
47150  2023-12-25/2023-12-31        4994                     4   

       weekly_avg_price  weekly_avg_discount_percent  total_revenue  
0                316.16                          0.0        1264.64  
1                 51.42                         30.0         107.97  
2                436.82                          0.0        2184.10  
3                402.62                         10.0        1087.08  
4                375.66                          5.0        1070.64  
...                 ...                          ...            ...  
47146            350.03                          0.0         350.03  
47147            439.64                         15.0         373.69  
47148            222.28                         20.0         355.64  
47149            131.93                         10.0         118.74  
47150            139.23                         15.0         473.40  

[47151 rows x 6 columns]

In [ ]:
weekly_df.shape

(47151, 6)

In [23]:
weekly_df.head()

year_week  product_id  weekly_quantity_sold  weekly_avg_price  \
0  2021-12-27/2022-01-02        1042                     4            316.16   
1  2021-12-27/2022-01-02        1050                     3             51.42   
2  2021-12-27/2022-01-02        1080                     5            436.82   
3  2021-12-27/2022-01-02        1094                     3            402.62   
4  2021-12-27/2022-01-02        1140                     3            375.66   

   weekly_avg_discount_percent  total_revenue  
0                          0.0        1264.64  
1                         30.0         107.97  
2                          0.0        2184.10  
3                         10.0        1087.08  
4                          5.0        1070.64

### Step 6 — Identify Top SKUs by Total Demand

In [24]:
top_skus = (
    weekly_df
    .groupby("product_id")["weekly_quantity_sold"]
    .sum()
    .sort_values(ascending=False)
    .head(15)
)

In [25]:
top_skus

product_id
1087    81
1580    80
3537    79
4217    79
1617    78
2048    78
1386    77
2785    77
4204    77
3700    75
4796    73
4264    72
3809    72
1266    72
4933    71
Name: weekly_quantity_sold, dtype: int64

filtering the dataset

In [26]:
weekly_top_df = (
    weekly_df[weekly_df['product_id'].isin(top_skus.index)]
).copy()

In [27]:
weekly_top_df['product_id'].nunique()

15

In [31]:
weekly_top_df.shape

(308, 6)

- If you have 105 weeks and 15 SKUs, ideal full matrix would be: 105 × 15 = 1575 rows.
- we only have 308.

That means:
- Many SKU–week combinations are missing.

Which means:
- Those SKUs did not sell in many weeks.

This is important.
- Right now out dataset only contains weeks where at least one transaction occurred.
- But forecasting requires continuous weekly time series.
- We must create missing weeks with zero demand.

### Step 7 — Create Complete Weekly Grid

In [33]:
# Extract unique weeks:
all_weeks = weekly_df["year_week"].unique()

In [34]:
# Extract top SKU list:
sku_list = top_skus.index

In [35]:
sku_list

Index([1087, 1580, 3537, 4217, 1617, 2048, 1386, 2785, 4204, 3700, 4796, 4264,
       3809, 1266, 4933],
      dtype='int64', name='product_id')

In [36]:
# Create full grid:
full_index = pd.MultiIndex.from_product(
    [all_weeks, sku_list],
    names=["year_week", "product_id"]
)

full_df = pd.DataFrame(index=full_index).reset_index()

In [37]:
full_df

year_week  product_id
0     2021-12-27/2022-01-02        1087
1     2021-12-27/2022-01-02        1580
2     2021-12-27/2022-01-02        3537
3     2021-12-27/2022-01-02        4217
4     2021-12-27/2022-01-02        1617
...                     ...         ...
1570  2023-12-25/2023-12-31        4796
1571  2023-12-25/2023-12-31        4264
1572  2023-12-25/2023-12-31        3809
1573  2023-12-25/2023-12-31        1266
1574  2023-12-25/2023-12-31        4933

[1575 rows x 2 columns]

In [ ]:
# Merge with your filtered data
weekly_top_df = full_df.merge(
    weekly_top_df,
    on=["year_week", "product_id"],
    how="left"
)

In [39]:
# Fill missing demand with 0

weekly_top_df["weekly_quantity_sold"] = weekly_top_df["weekly_quantity_sold"].fillna(0)

In [40]:
weekly_top_df.shape

(308, 6)

In [41]:
weekly_top_df["product_id"].nunique()

15

In [42]:
weekly_top_df["year_week"].nunique()

100